integration/overrides_exporter_test.go

```diff
@@ -0,0 +1,154 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+//go:build requires_docker
+// +build requires_docker
+
+package integration
+
+import (
+	"context"
+	"errors"
+	"path/filepath"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/grafana/dskit/backoff"
+	"github.com/grafana/e2e"
+	e2edb "github.com/grafana/e2e/db"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/stretchr/testify/require"
+
+	"github.com/grafana/mimir/integration/e2emimir"
+)
+
+const (
+	overridesFileName            = "overrides.yaml"
+	overridesFileContentTemplate = `
+overrides:
+  tenant-a:
+    ingestion_rate: {{.MaxIngestionRate}}
+`
+)
+
+func TestOverridesExporterTenantSharding(t *testing.T) {
+	limit := 20
+	tests := []struct {
+		name              string
+		flags             map[string]string
+		expectedMetricSum int
+	}{
+		{
+			name: "tenant sharding enabled",
+			flags: map[string]string{
+				"-overrides-exporter.ring.enabled": "true",
+			},
+			// the limit should only be exported once, i.e. the expected sum of metrics should equal the limit value
+			expectedMetricSum: limit,
+		},
+		{
+			name: "tenant sharding disabled",
+			flags: map[string]string{
+				"-overrides-exporter.ring.enabled": "false",
+			},
+			// the limit should be exported from both exporters, i.e. the expected sum is twice the limit value
+			expectedMetricSum: 2 * limit,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			s, err := e2e.NewScenario(networkName)
+			require.NoError(t, err)
+			defer s.Close()
+
+			require.NoError(t, writeFileToSharedDir(
+				s,
+				overridesFileName,
+				[]byte(buildConfigFromTemplate(overridesFileContentTemplate, map[string]int{"MaxIngestionRate": limit})),
+			))
+
+			flags := map[string]string{
+				"-runtime-config.file": filepath.Join(e2e.ContainerSharedDir, overridesFileName),
+			}
+
+			services, err := newOverridesExporterServices(mergeFlags(flags, tt.flags), s)
+			require.NoError(t, err)
+
+			require.NoError(t, waitTenantSpecificMetricsAvailable(services.e1, services.e2, "tenant-a"))
+
+			value1, err := getOverrideMetricForTenantFromService("tenant-a", "ingestion_rate", services.e1)
+			require.NoError(t, err)
+			value2, err := getOverrideMetricForTenantFromService("tenant-a", "ingestion_rate", services.e2)
+			require.NoError(t, err)
+
+			require.Equal(t, tt.expectedMetricSum, int(value1+value2))
+		})
+	}
+}
+
+func waitTenantSpecificMetricsAvailable(s1, s2 *e2emimir.MimirService, tenantName string) error {
+	ctx, cancel := context.WithTimeout(context.Background(), 20*time.Second)
```

Comment:
[nit] I think this code would be easier if you will use the `test.Poll()` utility.